### PR TITLE
[exporter/datadog] Use statswriter, client aggregator is for tracer stats only

### PR DESCRIPTION
**Pull request by ajgajg1134**
**Description:** <Describe what has changed.>

The `ProcessStats` function receives "tracer stats" payloads, by looping through the payload and breaking it apart this drastically increases the number of stats payloads being emitted. Instead, let's write directly to the StatsWriter, which reduces duplicated work and keeps the payload together as a single payload.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>
I tested this locally with the otel example calendar app verifying stats were calculated correctly as well, the logs also showed the reduction in the number of payloads being emitted.

~TODO: I will add a test to prevent future regression~ Done!

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by songy23 — COMMENTED** [review]
Can you add a changelog? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/.chloggen/TEMPLATE.yaml

File: exporter/datadogexporter/factory.go [hunk 1]
```diff
@@ -283,11 +283,20 @@ func (f *factory) createMetricsExporter(
 		cancel()
 		return nil, fmt.Errorf("failed to start trace-agent: %w", err)
 	}
-	var statsOut chan []byte
+	acfg, err := newTraceAgentConfig(ctx, set, cfg, hostProvider)
+	if err != nil {
+		return nil, err
```
Comment (songy23) [inline exporter/datadogexporter/factory.go:288]: Need to call `cancel()` before this line
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/7849700829/job/21424234053?pr=31173

File: exporter/datadogexporter/factory.go [hunk 2]
```diff
@@ -283,11 +283,21 @@ func (f *factory) createMetricsExporter(
...
+		cancel()
+		return nil, err
+	}
+	statsToAgent := make(chan *pb.StatsPayload)
+	statsWriter := writer.NewStatsWriter(acfg, statsToAgent, telemetry.NewNoopCollector())
+
+	set.Logger.Debug("Starting Datadog Trace-Agent StatsWriter")
+	go statsWriter.Run()
+
+	var statsIn chan []byte
```
Comment (songy23) [inline exporter/datadogexporter/factory.go:297]: Shouldn't this block be in the if condition below?

File: exporter/datadogexporter/factory.go [hunk 3]
```diff
@@ -237,7 +239,7 @@ func checkAndCastConfig(c component.Config, logger *zap.Logger) *Config {
 	return cfg
 }
 
-func (f *factory) consumeStatsPayload(ctx context.Context, out chan []byte, traceagent *agent.Agent, tracerVersion string, logger *zap.Logger) {
+func (f *factory) consumeStatsPayload(ctx context.Context, statsIn <-chan []byte, statsToAgent chan<- *pb.StatsPayload, logger *zap.Logger) {
```
Comment (mackjmr) [inline exporter/datadogexporter/factory.go:242]: This only get's called when the `ConnectorPerformanceFeatureGate` is enabled. This is going to be the default at some point, but for now it isn't and I believe we also need make the change [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2bf56cacf1428e13483f2dea5ef237475f3ca48f/exporter/datadogexporter/metrics_exporter.go#L263-L265)

File: exporter/datadogexporter/factory.go [hunk 4]
```diff
@@ -283,11 +283,21 @@ func (f *factory) createMetricsExporter(
...
 		return nil, fmt.Errorf("failed to start trace-agent: %w", err)
 	}
-	var statsOut chan []byte
+	acfg, err := newTraceAgentConfig(ctx, set, cfg, hostProvider)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	statsToAgent := make(chan *pb.StatsPayload)
+	statsWriter := writer.NewStatsWriter(acfg, statsToAgent, telemetry.NewNoopCollector())
```
Comment (jdgumz) [inline exporter/datadogexporter/factory.go:292]: Why a noop collector? Is it because statsd submission won't work as expected here?

**Review by jdgumz — APPROVED** [review]
✅ the fundamental change looks right to me

File: exporter/datadogexporter/factory.go (hunk 4 above)
Comment (ajgajg1134) [reply exporter/datadogexporter/factory.go:292]: Yeah I believe that's why? Although I was just copying the original existing behavior which uses a noop collector in `newTraceAgent`

**Comment by songy23** [comment]
Please fix lint https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/7878438470/job/21496937483?pr=31173#step:8:333

**Review by dineshg13 — COMMENTED** [review]
The statswriter code isn't setting the tracer version as done for clientstatsAggregator(https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/378b552d255040a73a81e9090cca80a42a65db5d/exporter/datadogexporter/metrics_exporter.go#L262) . Can you please validate if it is necessary / what its value is with the PR.

**Comment by ajgajg1134** [comment]
> The statswriter code isn't setting the tracer version as done for clientstatsAggregator(
> 
> https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/378b552d255040a73a81e9090cca80a42a65db5d/exporter/datadogexporter/metrics_exporter.go#L262
> 
> ) . Can you please validate if it is necessary / what its value is with the PR.

It looks like tracer version is just an "informative" field, I ran it locally and did verify that the new code gave an empty string. So to keep the previous behavior I just added a commit that will attach the tracer version to the client stats payload if it's not already there

**Review by songy23 — COMMENTED** [review]
Toolchain is having issue right now https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31213. Should be fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31214

**Review by knusbaum — APPROVED** [review]
Looks great.

**Review by mx-psi — APPROVED** [review]

**Pull request merged**